### PR TITLE
Hero UI stats

### PR DESCRIPTION
**Pull request by Surxe**
Hero's UI attribute stats parsed, plan is to use it to determine which stats should go in the hero infobox

### CODE REVIEW

**Comment by Surxe** [comment]
Actually, going to add shop ui stats as well before merging

**Comment by Surxe** [comment]
> Actually, going to add shop ui stats as well before merging

done, rdy for review

File: src/parser/parsers/attributes.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+    def run(self):
+        all_attributes = {}
+        all_attributes = self._parse_stats_ui(self.heroes_data)
+
+        json_utils.write(
+            OUTPUT_DIR + 'json/attribute-data.json', json_utils.sort_dict(all_attributes)
+        )
+
+    # Move StatsUI data to attributes data
```
Comment (hassayag) [inline src/parser/parsers/attributes.py:24]: Minor, but I think any docs for a function should follow this convention, as it will show the comment when you hover over any use of the function. 
```python
    def _parse_stats_ui(self, heroes_data):
        """Move StatsUI data to attributes data"""
```

File: src/parser/parsers/attributes.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            if 'StatsUI' not in hero_value:
+                continue
+
+            stats_ui = hero_value['StatsUI']
+            for ui_category, stats in stats_ui.items():
+                if ui_category not in all_attributes:
+                    all_attributes[ui_category] = []
+
+                for stat_name in stats:
+
```
Comment (hassayag) [inline src/parser/parsers/attributes.py:38]: empty line not needed

File: src/parser/parsers/heroes.py [hunk 3]
```diff
@@ -93,6 +109,108 @@ def _parse_hero_weapon(self, hero_value):
...
+            "ReloadSpeed": "Weapon",
+            "BulletSpeed": "Weapon",
+            "BulletSpeedIncrease": "Weapon",
+            "BulletLifesteal": "Weapon",
+            "LightMeleeDamage": "Weapon",
+            "HeavyMeleeDamage": "Weapon"
+            "WeaponAttribute_RapidFire": "Weapon"
+            "WeaponAttribute_MediumRange": "Weapon"
+            
+        """
```
Comment (hassayag) [inline src/parser/parsers/heroes.py:171]: I'm not sure the logic is complex enough to warrant such a wall of text. 
Comments are fine, but place them above their relevant block of code and keep it concise

File: src/utils/json_utils.py [hunk 4]
```diff
@@ -13,6 +13,23 @@ def write(path, data):
         json.dump(data, outfile, indent=4)
 
 
+# Remove keys from a dictionary at the specified depth
+# Depth 1 removes keys from the first level of the dictionary
+# Depth N removes keys from the first N levels of the dictionary
+def remove_keys(data, keys_to_remove, depths_to_search=1):
```
Comment (hassayag) [inline src/utils/json_utils.py:19]: similar to my first comment, document the function such that the intellisense will show it along with the function

File: src/parser/parsers/heroes.py [hunk 5]
```diff
@@ -55,13 +55,29 @@ def run(self):
...
+                        raise Exception(
+                            f'Hero {hero_key} has {mult_str} of {mult_value} instead of 1'
+                        )
 
-        json_utils.write(OUTPUT_DIR + 'json/hero-data.json', json_utils.sort_dict(all_hero_stats))
+        # Include removed keys in the data sent to consecutive parsers, but not to the output file
+        hero_stats_to_remove = multipliers + ['']
+        hero_stats_removed = json_utils.remove_keys(
+            all_hero_stats, keys_to_remove=hero_stats_to_remove, depths_to_search=2
+        )
```
Comment (hassayag) [inline src/parser/parsers/heroes.py:76]: why do we want to remove multipliers?

File: src/parser/parsers/heroes.py [hunk 6]
```diff
@@ -32,7 +32,7 @@ def run(self):
                 hero_stats.update(
                     self._map_attr_names(hero_value['m_mapStartingStats'], maps.get_hero_attr)
                 )
-
+                hero_stats['StatsUI'] = self._parse_stats_ui(hero_value)
```
Comment (hassayag) [inline src/parser/parsers/heroes.py:35]: Firstly, you have not run the parser, which is why this change is not reflected on in the output-data.
Secondly, this appends the exact same data to every hero object. Is this needed?

File: src/parser/parsers/attributes.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    def run(self):
+        all_attributes = {}
+        all_attributes = self._parse_stats_ui(self.heroes_data)
+
+        json_utils.write(
+            OUTPUT_DIR + 'json/attribute-data.json', json_utils.sort_dict(all_attributes)
+        )
+
+    # Move StatsUI data to attributes data
+    def _parse_stats_ui(self, heroes_data):
```
Comment (hassayag) [inline src/parser/parsers/attributes.py:25]: This file is all about parsing the attributes, but the actual parsing is happening in heroes.py? Shouldn't it be the other way round, such that we
1. parse attrs
2. pass parsed attrs to heroes parser (if needed)

Or are there complications with that method?

**Review by hassayag — CHANGES_REQUESTED** [review]
Cool to have a definite list of all the UI attrs. I just have concerns for the overall structuring

Also remember to run the parser with the latest game files.

PS. gonna start using `feature/*` branch names, along with `bugfix` and `hotfix` as well. So just a note for any future branches

File: src/parser/parsers/attributes.py (hunk 1 above)
Comment (Surxe) [reply src/parser/parsers/attributes.py]: Good point, will start doing this. Completed.

File: src/parser/parsers/attributes.py (hunk 2 above)
Comment (Surxe) [reply src/parser/parsers/attributes.py]: complete

File: src/parser/parsers/heroes.py (hunk 3 above)
Comment (Surxe) [reply src/parser/parsers/heroes.py]: Removed the excessive examples; its now as short as possible

File: src/utils/json_utils.py (hunk 4 above)
Comment (Surxe) [reply src/utils/json_utils.py:19]: complete

File: src/parser/parsers/heroes.py (hunk 5 above)
Comment (Surxe) [reply src/parser/parsers/heroes.py:76]: See above; these are specifically stats that are constants for all heroes and as a result are not listed in the hero infobox. With upcoming infobox changes, this will definitely be moved/changed anyways.

File: src/parser/parsers/heroes.py (hunk 6 above)
Comment (Surxe) [reply src/parser/parsers/heroes.py]: completely reworked

File: src/parser/parsers/attributes.py (hunk 7 above)
Comment (Surxe) [reply src/parser/parsers/attributes.py]: discussed externally; completely reworked.

**Review by hassayag — APPROVED** [review]
Looks good! Will rerun the parser on master to get the latest data
